fib: changed handling of the net prefix by the FIB

BytesGalore opened the pull request:
Until now the prefix length for a given destination has been determined automatically by the FIB.
This PR changes it to be provided as upper byte in the `global_flags` of an entry.
Inspired by the IP scheme, the number of significant bits is stored in the designated most upper byte of the flags.

Rationale:
Before this change there was for instance no way to set a `>116` prefix for a private subnet like `2001::1234:0000`.
The API CHANGE Label belongs only to the reserved upper byte of the `global_flags` of a FIB entry used as storage for the prefix bits.

cgundogan commented on sys/net/gnrc/routing/rpl/gnrc_rpl_dodag.c:
```diff
@@ -243,7 +243,7 @@ void gnrc_rpl_parent_update(gnrc_rpl_dodag_t *dodag, gnrc_rpl_parent_t *parent)
             if ((if_id = gnrc_ipv6_netif_find_by_addr(NULL, &all_RPL_nodes)) != KERNEL_PID_UNDEF) {
                 fib_add_entry(&gnrc_ipv6_fib_table, if_id, def.u8, sizeof(ipv6_addr_t),
-                              (FIB_FLAG_NET_PREFIX | 0x0),
+                              (0x00 << FIB_FLAG_NET_PREFIX_SHIFT | 0x0),
```
`0x00 << FIB_FLAG_NET_PREFIX_SHIFT` == 0`?

BytesGalore replied:
yup, made it to keep all cases equal.
`(0x00 << FIB_FLAG_NET_PREFIX_SHIFT)` flag for `IPV6_ADDR_UNSPECIFIED` is now considered as default route by the FIB (inspired by `::/0` notation).

cgundogan replied:
Why don't you just write `0x0` instead of `0x00 << FIB_FLAG_NET_PREFIX_SHIFT | 0x0` ?

BytesGalore replied:
I changed it :)

cgundogan commented:
I am jts wondering: aren't 7 bits sufficient for storing the prefix?

BytesGalore commented:
For IPv6 yes.
But I aim the FIB to be more flexible then just support up to IPv6 :)

cgundogan commented on sys/net/gnrc/routing/rpl/gnrc_rpl_dodag.c:
```diff
@@ -243,7 +243,7 @@ void gnrc_rpl_parent_update(gnrc_rpl_dodag_t *dodag, gnrc_rpl_parent_t *parent)
             if ((if_id = gnrc_ipv6_netif_find_by_addr(NULL, &all_RPL_nodes)) != KERNEL_PID_UNDEF) {
                 fib_add_entry(&gnrc_ipv6_fib_table, if_id, def.u8, sizeof(ipv6_addr_t),
-                              (FIB_FLAG_NET_PREFIX | 0x0),
+                              0x0,
```
you also need to adjust line 308 otherwise it won't compile

BytesGalore replied:
done

cgundogan commented:
```
RIOT/sys/net/network_layer/fib/fib.c:1535:28: error: format '%d' expects argument
                     of type 'int', but argument 2 has type 'uint32_t
                     {aka long unsigned int}' [-Werror=format=]
                     printf("N /%d ", (prefix >> FIB_FLAG_NET_PREFIX_SHIFT));
```

BytesGalore commented:
arrrgh, done.

BytesGalore commented:
rebased

cgundogan commented:
I am testing this change again

Yonezawa-T2 commented on sys/shell/commands/sc_fib.c:
```diff
@@ -74,15 +74,33 @@ static void _fib_usage(int info)
 {
-    unsigned char *dst = (unsigned char *)dest;
-    size_t dst_size = (strlen(dest));
+    uint32_t prefix = 0;
```
If prefix is omitted, should it be 0 or 128?

BytesGalore replied:
The FIB omits the prefix if the specified byte in the falg is 0 **or** the length of the address in bits.
So 0 omits the prefix for any used address type, e.g. IPv6, IPv4, ..., and 128 omits an IPv6 prefix.

cgundogan commented:
Sorry for delaying this so much. I will be able to look at it again at the end of next week.

BytesGalore commented:
@cgundogan np :) , the release is/was way more important than this PR.

cgundogan commented:
@BytesGalore could you rebase again and adjust the changes that were added in #4661? (The compiler complains, because `FIB_FLAG_NET_PREFIX` does not exist anymore)

BytesGalore commented:
ok, Im on it

BytesGalore commented:
done.

cgundogan commented:
the output of fibroute seems to be a little bit off (indentation-wise)

BytesGalore commented:
@cgundogan you mean the large distance between IP and flags right?

cgundogan commented:
On the testbed:
```
1453833644.340535;m3-11;> fibroute
1453833644.344417;m3-11;Destination                             Flags        Next Hop                                Flags      Expires          Interface
1453833644.346350;m3-11;abcd::                                  0x40000000 N /64 fe80::3432:4833:46d4:962a               0x00000000 NEVER            7
```

cgundogan commented:
@BytesGalore any news here? (;

BytesGalore commented:
oh, nope.
This one somehow gone under, will have a look this afternoon.

BytesGalore commented:
Now news :)
should look better now

cgundogan commented:
there were a lot of discussions regarding the fib and the border router in the last days. I'd like to ask @kaspar030 @OlegHahm or @authmillenon to have a quick look at this PR also to not break something that they are working on right now.

cgundogan commented:
@BytesGalore can you rebase?

cgundogan commented:
and also squash

OlegHahm commented:
I promise to take a look a next H&A.

cgundogan commented:
`FIB_FLAG_NET_PREFIX` undeclared in `gnrc_rpl_dodag.c:207`

BytesGalore commented:
@cgundogan I replaced `FIB_FLAG_NET_PREFIX` and directly squashed

BytesGalore commented:
@cgundogan nice thx for reviewing :)

OlegHahm commented on sys/universal_address/universal_address.c:
```diff
@@ -196,41 +196,47 @@ int universal_address_compare(universal_address_container_t *entry,
+    /* compare up to fist distinct byte, pretty clumsy method for now */
+    int idx = -1;
+    bool test_all_zeros = true;
+    for( size_t i = 0; i < entry->address_size; i++) {
```
superfluous white-space before `size_t`.

OlegHahm replied:
here and below

OlegHahm commented on sys/universal_address/universal_address.c:
```diff
@@ -196,41 +196,47 @@ int universal_address_compare(universal_address_container_t *entry,
+    if ( test_all_zeros ) {
+        *addr_size_in_bits = 0;
+        mutex_unlock(&mtx_access);
+        return 2;
```
I would prefer constant instead of a literal here (and the doxygen needs updating)

OlegHahm commented on sys/universal_address/universal_address.c:
```diff
@@ -196,41 +196,47 @@ int universal_address_compare(universal_address_container_t *entry,
+
     mutex_unlock(&mtx_access);
     return ret;
 }
```
@cgundogan thinks that can be simplified - he will open an issue.

OlegHahm commented on sys/net/network_layer/fib/fib.c:
```diff
@@ -132,8 +132,7 @@ static int fib_find_entry(fib_table_t *table, uint8_t *dst, size_t dst_size,
             /* If we found an exact match */
-            if (ret_comp == 0 || (is_all_zeros_addr && match_size == 0)) {
-                DEBUG("[fib_find_entry] found an exact match");
+            if (ret_comp == 0 || (is_all_zeros_addr && (match_size == 0) && (ret_comp == 2))) {
```
add parens around single terms

OlegHahm replied:
Isn't `is_all_zeros_addr` and `(ret_comp == 2)` redundant?

PeterSchmerzl commented:
ACK

OlegHahm commented on sys/net/network_layer/fib/fib.c:
```diff
@@ -141,17 +140,35 @@ static int fib_find_entry(fib_table_t *table, uint8_t *dst, size_t dst_size,
+                                                    & FIB_FLAG_NET_PREFIX_MASK) >> FIB_FLAG_NET_PREFIX_SHIFT;
+
+                        if ((match_size >= global_prefix_len) &&
+                            ((prefix_size == 0) || (match_size > prefix_size))) {
```
`prefix_size` is initialized as 0 and not set anywhere before to something else.

OlegHahm commented:
ACK - apart from minor comments.

BytesGalore commented on sys/universal_address/universal_address.c:
```diff
@@ -248,31 +254,31 @@ int universal_address_compare_prefix(universal_address_container_t *entry,
-                for(; i < entry->address_size; ++i) {
-                    if( entry->address[i] != 0 ) {
-                        ret = 1;
+                for ( ; i < entry->address_size; ++i) {
```
I intentionally added the whitespace before the first `;` in the `for` statement because I tend to overlook them in such cases.

OlegHahm replied:
Yeah, that's okay here.

BytesGalore commented:
@OlegHahm I guess I addressed all your comments, thx for reviewing :)
I made a separate companion commit for the doxy/define changes in `universal_address`

@PeterSchmerzl welcome on the code side of the force, and thx for your ACK

OlegHahm commented:
Murdock has some complaints.

BytesGalore commented:
ok, I think I got 'em

cgundogan commented on sys/include/universal_address.h:
```diff
@@ -107,9 +118,12 @@ uint8_t* universal_address_get_address(universal_address_container_t *entry,
- *         -ENOENT if the given adresses do not match
+ * @return UNIVERSAL_ADDRESS_EQUAL if the entries are equal
+ * @return UNIVERSAL_ADDRESS_MATCHING_PREFIX if the entry matches to a certain prefix 
+ *                                           (trailing '0's in *entry)
```
s/*entry/@p entry/

cgundogan commented on sys/include/universal_address.h:
```diff
@@ -128,9 +142,10 @@ int universal_address_compare(universal_address_container_t *entry,
-*         -ENOENT if the given adresses do not match
+* @return UNIVERSAL_ADDRESS_EQUAL if the entries are equal
+* @return UNIVERSAL_ADDRESS_MATCHING_PREFIX if the entry matches to a certain prefix 
+*                                           (trailing '0's in *prefix)
```
s/*entry/@p entry/

cgundogan commented on sys/include/universal_address.h:
```diff
@@ -107,9 +118,12 @@ uint8_t* universal_address_get_address(universal_address_container_t *entry,
- *         -ENOENT if the given adresses do not match
+ * @return UNIVERSAL_ADDRESS_EQUAL if the entries are equal
+ * @return UNIVERSAL_ADDRESS_MATCHING_PREFIX if the entry matches to a certain prefix
+ *                                           (trailing '0's in *entry)
```
s/*entry/@p entry/

BytesGalore replied:
done

cgundogan commented on sys/include/universal_address.h:
```diff
@@ -128,9 +142,10 @@ int universal_address_compare(universal_address_container_t *entry,
-*         -ENOENT if the given adresses do not match
+* @return UNIVERSAL_ADDRESS_EQUAL if the entries are equal
+* @return UNIVERSAL_ADDRESS_MATCHING_PREFIX if the entry matches to a certain prefix
+*                                           (trailing '0's in *prefix)
```
s/*entry/@p prefix/

BytesGalore replied:
done and directly squashed

cgundogan commented:
Okay, I did a last test on the iotlab testbed with routing over 3 hops. Worked fine, my ACK upholds, @OlegHahm has already acked and Murdock is green => GO

cgundogan commented:
Congrats!

BytesGalore commented:
nice thx again :D
